### PR TITLE
systemd unit file

### PR DESCRIPTION
**Pull request by klingtnet**
This pull requests adds a systemd unit file for caddy and a bit of documentation on how to install and use caddy as a systemd service. Refers to #388

### CODE REVIEW

**Comment by abiosoft** [comment]
Thanks for this :smile: Users no longer need to look far for systemd config.

Can you squash the commits into one ?

**Comment by klingtnet** [comment]
@abiosoft For sure! The commit squashing should be mentioned in the [CONTRIBUTORS.md](https://github.com/mholt/caddy/blob/master/CONTRIBUTING.md).

**Comment by mholt** [comment]
Thanks for this!

Does this follow a PID file? What if the user signals Caddy with USR1? Also, what is the value of $HOME in the execution environment? (In other words, where does Caddy write the .caddy folder?)

**Comment by klingtnet** [comment]
The `.caddy` folder will be written into the `home` directory of the user that runs caddy, which is `someuser` for `systemctl start caddy@someuser`. Maybe I should add this to the README.
Systemd keeps track of the processes PID, but if one needs to know it he can use: `systemctl show --property=MainPID caddy@someuser.service`.
Sending signals is straight-forward: `systemctl kill --signal=USR1 caddy@someuser`

**Comment by klingtnet** [comment]
I've updated the README.md with the signal instructions and added a note on where the `.caddy` folder is created.

File: dist/init/linux-systemd/caddy@.service [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+[Service]
+; run user and group for caddy
+User=%i
+Group=http
+ExecStart=/usr/bin/caddy -agree=true -conf=/etc/caddy/Caddyfile
+Restart=on-failure
+; create a private temp folder that is not shared with other processes
+PrivateTmp=true
+; limit the number of file descriptors, see `man systemd.exec` for more limit settings
+LimitNOFILE=8192
```
Comment (mholt) [inline dist/init/linux-systemd/caddy@.service]: Can this go higher? It should be cranked as high as possible for production sites. (I admit the recommendation that we print about ulimit -n is too low.)

**Comment by klingtnet** [comment]
I don't know if this is totally true, but the number of open file descriptors is mainly limited by the systems memory and maybe some other limitations set by the OS. Nevertheless, we could easily increase this to something higher as `8192`. I'm assuming that caddy closes the file descriptors after usage and does not leak them :grin:.

**Comment by klingtnet** [comment]
I studied the file descriptor topic a bit and found the following limits:
- `cat /proc/sys/fs/file-max` prints the maximum number of file-handles that the installed kernel will allocate (http://www.tldp.org/LDP/solrhe/Securing-Optimizing-Linux-RH-Edition-v1.3/chap6sec72.html). These are `806198` on my laptop (stock Arch Linux kernel) and `609124` on my KVM based vServer.
- `/etc/security/limits.conf` can contain soft and hard limits for file-handles users and groups

The file-handle limit very much depends on the use-case of caddy, if one uses it solely as a reverse proxy then he will need a lot less handles than if it serves a lot of static content. I don't have a recommendation but we should be safe using something > 10000, though.

**Comment by mholt** [comment]
It doesn't leak fds (as far as we know) but still, a very busy site -- especially one that is proxying -- is at risk of running out of file descriptors. Setting it low in production doesn't have much utility anyway unless you're a very savvy sysadmin trying to balance several busy apps on the same machine. Prefer to set closer to 16384. (That's what I do!)

**Comment by mholt** [comment]
No reason not to merge this now I guess. @klingtnet I'll give you push access so you can maintain your init script. Also feel free to review and help out with other issues and PRs as much as you want!

**Pull request merged**